Circular Convolution Function via circular padding

NarineK opened the pull request:
Summary: Added circular padding in addition to zero padding to Conv1D, Conv2D and Conv3D based on the solution suggested in: https://github.com/pytorch/pytorch/issues/3858

Differential Revision: D14126416

ezyang commented on test/common_nn.py:
```diff
@@ -2490,6 +2535,60 @@ def ctcloss_reference(log_probs, targets, input_lengths, target_lengths, blank=0
+
+def padding1d_circular(input):
+    return torch.cat([input[:, :, -1:], input,
+                      input[:, :, 0:1]], dim=2)
```
In fact, your reference implementations seem to only implement the pad=1 case. Perhaps this was intentional? But it doesn't seem like we're getting very much utility from the reference implementation in this case.

One possible alternative is to use the implementation that looked like `torch.cat(input, input, input)` and then view out the bits we didn't need. Or we could just drop the reference implementation entirely; I agree it doesn't seem necessary in this case.

NarineK replied:
I can add more test cases with and without reference implementation

ezyang commented on torch/nn/functional.py:
```diff
@@ -33,8 +33,9 @@
     stride: the stride of the convolving kernel. Can be a single number or
       a one-element tuple `(sW,)`. Default: 1
-    padding: implicit zero paddings on both sides of the input. Can be a
+    padding: implicit zero or circular paddings on both sides of the input. Can be a
```
nit: I'd probably just say "implicit paddings" here and let `padding_mode` explain what the input padding is. Otherwise you have to keep updating this list whenever people add more padding modes.

ezyang commented on torch/nn/functional.py:
```diff
@@ -33,8 +33,9 @@
-    padding: implicit zero paddings on both sides of the input. Can be a
+    padding: implicit zero or circular paddings on both sides of the input. Can be a
       single number or a one-element tuple `(padW,)`. Default: 0
+    padding_mode: input paddings accepts values `zeros` or `circular`. Default: `zeros`
```
nit: "input paddings accepts"? How about "what kind of padding is applied to both sides. Can be"

NarineK replied:
Oh, I think that is a typo. I meant the type of padding `circular` or `zeros`. I'll fix it. Thank you for catching

ezyang commented on torch/nn/functional.py:
```diff
@@ -2984,3 +2987,28 @@ def fold(input, output_size, kernel_size, dilation=1, padding=0, stride=1):
+    Returns
+        :return: tensor of shape :math:`(N, C_{\text{in}}, [D + 2 * padding[0],
+                 H + 2 * padding[1]], W + 2 * padding[2]))`
+    """
```
After reading this implementation, I realized (this was NOT in the original bootcamp task) that we can get error checking (which is conspicuously missing here) by folding `pad_circular` into `torch.nn.functional.pad`, which has a number of appropriate checks. It's probably a good idea to make sure these implementations line up anyway. Would you mind doing this? Thanks (and sorry for not realizing that was a better way to do it to start with.)

NarineK replied:
I actually read through the implementation of pad(...). I think it is a good idea to combine those two functions. I'll work on it.

NarineK replied:
Hi @ezyang, I looked into the implementation of pad() and it looks that the padding requirements are slightly different there. E.g. it requires that the length of pad is even: `assert len(pad) % 2 == 0, 'Padding length must be divisible by 2'`

NarineK replied:
Do we  still want to fold `pad_circular` under `pad`, @ezyang ? 
For Conv3D padding can look something like this: 
```
m = nn.Conv3d(16, 33, (3, 5, 2), stride=(2, 1, 1), padding=(4, 2, 0))
```
This will fail the checks. It looks like in pad(...) function paddings are expected to have different format.

soumith replied:
F.pad wants its `pad` to be divisible by 2, to cover two sides of a dimension.
For example, (left_pad, right_pad), or (left_pad, right_pad, top_pad, bottom_pad), etc.

So, you should make `circular` a mode in `F.pad`, and in Conv*d, translate the padding to F.pad's equivalent.

For example, 

```
m = nn.Conv3d(16, 33, (3, 5, 2), stride=(2, 1, 1), padding=(4, 2, 0))
```
should end up calling (internally to Conv3d) `F.pad(input, (0, 0, 1, 1, 2, 2), mode='circular')`.

Note carefully that in Conv3d, the padding is given from left dimension to right dimension, i.e. `4` is the padding for D, `2` for H and `0` for W dimension of the input of dimensions N x C x D x H x W. And in `F.pad`, one gives padding dimension from right to left, so you give `(W_left, W_right, H_left, H_right, D_left, D_right)`.

ezyang replied:
Hi @NarineK; sorry about the late response. It looks like `torch.nn.functional.pad` supports asymmetric padding (which is not directly supported by convolution). So, to convert a convolution style padding into a pad style padding, you would have to write `(4, 4, 2, 2, 0, 0)`.

soumith replied:
@ezyang see my comment above. it's not (4, 4, 2, 2, 0, 0), my comment gives the right ordering, reasoning and values.

ezyang replied:
Somehow, your comment didn't show up at all when I was looking.

NarineK replied:
I see, this will change the logic for the padding. In case of padding=(5, 2, 0)). I'll have something like: ```F.pad(input, (0, 0, 1, 1, 2, 3)```

NarineK replied:
Pushed the changes

NarineK replied:
addressed

ezyang reviewed (changes requested):
This is a really good start. I think the big things for me are:

1. Getting better test coverage for padding > 1 case. I don't really care if you have a reference implementation, but we should exercise it.
2. Fixing the error checking in `circular_pad`. My suggestion is to fold it into the existing `pad` function.

apaszke reviewed (commented):
Could we just have `F.pad_circular` to avoid having a blow up for every single dimensionality?

NarineK commented on torch/nn/modules/conv.py:
```diff
@@ -307,17 +314,23 @@ class Conv2d(_ConvNd):
         stride = _pair(stride)
         padding = _pair(padding)
         dilation = _pair(dilation)
+        self.padding_mode = padding_mode
```
I originally had padding_mode in `ConvNd` but then I noticed that I have to also override it in other child nodes such as `ConvTranspose2d` and that variable will not be used in those child nodes. Is it still okay to have `padding_mode` in parent knowing that I have to have it in all child nodes ?

ezyang replied:
Oh I see. Hmm, I guess the problem is, we were technically supposed to add this padding support for ConvTranspose, but we haven't done it at this point. How about adding an error message for those child classes saying that non-zero padding is not supported? Alternately, if you want to, figure out how padding is supposed to work on transposed convolution and implement it.

NarineK commented on test/common_nn.py:
```diff
@@ -2249,6 +2249,51 @@ def fractional_max_pool3d_test(test_case):
+    ),
+    dict(
+        fullname='Padding3d_circular',
+        constructor=wrap_functional(F.pad_circular, padding=(0, 0, 1,)),
```
I can add more test cases for larger padding sizes.

ezyang commented:
@apaszke I don't know how to write a generalized circular pad that works for arbitrary dimensionality, without writing the kernel out (which may eventually be a good idea, but I wouldn't block this PR on it.) Any ideas? :)

apaszke commented:
I haven't actually thought about that, but it seems like the implementations for 1d and 3d look the same modulo the number of colons (which are equivalent to `slice(None)`), so we should be able to do this programmatically. Also, it would be good to have better examples in the docstrings, because they don't fully explain the semantics I think. Is it correct to say that in the 2D case we add the last row at the top, the first row at the bottom, the left column on the right, and the right column on the left? What happens with the indices that are not covered by those elements?

NarineK commented:
Thank you @apaszke and @ezyang ! I can play with the slice(None) to find a more elegant implementation for the dimensionality.
With respect to `What happens with the indices that are not covered by those elements?` The other elements do not change or effect anything. Here is an example from https://github.com/pytorch/pytorch/issues/3858 for padding=1
Before circular padding
[[0 1 2]
[3 4 5]
[6 7 8]]

After circular padding
[[8 6 7 8 6]
[2 0 1 2 0]
[5 3 4 5 3]
[8 6 7 8 6]
[2 0 1 2 0]]

NarineK commented:
@ezyang and @apaszke , I've just pushed the implementation of pad_circular which works for arbitrary dimensions. 
https://github.com/pytorch/pytorch/commit/6832149459c698ca2460670c6eb46e5948262e49#diff-c66288b9ce36978f377a1a20d32ec53dR2993
The common_nn.py tests passed but test_jit does not like input[slice(None)] in general.
I will address other comments in a separate commit.

NarineK commented:
After talking to jit team we came to the conclusion that currently there is no support for slice. They opened the following feature to work on: https://github.com/pytorch/pytorch/issues/17389

I'll revert back my changes and update docstring examples

NarineK commented on test/common_nn.py:
```diff
@@ -2249,6 +2249,52 @@ def fractional_max_pool3d_test(test_case):
+        module_name='Conv1d',
+        constructor_args=(4, 5, 3, 1, 1, 1, 1, True, 'circular'),
+        input_size=(2, 4, 10),
+        cudnn=True,
```
It seems that there is an issue on jit side. Although my input is 3 dimensional here but the input to forward(...) method is 5 dimensional.
I found it out as the jit test was suspiciously working for Conv3D and failing for Conv1D and Conv2D.
cc: @ezyang

soumith replied:
@suo any ideas on why this would happen?

ezyang replied:
It also looks like some other TorchScript tests are failing, but they really should not be.

NarineK replied:
TorchScript tests related to this pull request, @ezyang  ? On phabricator it doesn't show anything failing

ezyang commented:
@pytorchbot rebase this please

ezyang commented on test/common_nn.py:
```diff
@@ -2501,6 +2601,45 @@ def ctcloss_reference(log_probs, targets, input_lengths, target_lengths, blank=0
+    output: [[[[[2, 0, 1, 2, 0]]]]]
+    """
+    return torch.cat([input[:, :, :, :, -i:], input,
+                      input[:, :, :, :, 0:i]], dim=4)
```
This looks wrong. But there is some fishy business about all of these reference implementations. See above.

NarineK replied:
It didn't effect specific examples that I had but in general you're right. It has to cut dim=2 and dim=3. I will add it.

NarineK replied:
addressed

NarineK commented on torch/nn/modules/conv.py:
```diff
@@ -436,25 +454,34 @@ class Conv3d(_ConvNd):
     @weak_script_method
     def forward(self, input):
+        if self.padding_mode == 'circular':
+            expanded_padding = ((self.padding[2] + 1) // 2, self.padding[2] // 2,
```
We could also have a separate method to split left and right paddings.

NarineK commented:
@pytorchbot rebase this please

NarineK commented:
Do you guys have any other comments on this PR ?

NarineK committed:
Circular Convolution Function via circular padding (#17240)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/17240

Added circular padding in addition to zero padding to Conv1D, Conv2D and Conv3D based on the solution suggested in: https://github.com/pytorch/pytorch/issues/3858

Differential Revision: D14126416

fbshipit-source-id: 72cdb69b8c22e33b0ba9ae515c4d119f69ea45d9

fmassa commented on torch/nn/functional.py:
```diff
@@ -3004,3 +3014,29 @@ def fold(input, output_size, kernel_size, dilation=1, padding=0, stride=1):
+
+
+@weak_script
+def pad_circular(input, padding):
```
question: why didn't we add this functionality to `F.pad`?

NarineK replied:
Hi @fmassa! That was a suggestion to have padding functionally at one place.
cc: @soumith , @ezyang

fmassa commented on torch/nn/modules/conv.py:
```diff
@@ -173,17 +176,23 @@ class Conv1d(_ConvNd):
+            return F.conv1d(F.pad(input, expanded_padding, mode='circular'),
+                            self.weight, self.bias, self.stride,
+                            _single(0), self.dilation, self.groups)
         return F.conv1d(input, self.weight, self.bias, self.stride,
```
do we want to have the functional interface of `conv` also support the padding mode?

jzazo replied:
In the documentation it appears padding_mode is supported, but when I try to use it tells me it is an unexpected argument. I will do the manual padding, as in the module, but it would be convenient if the operation is supported.
In my case, my filter is a tensor, not a parameter, and I would prefer not to use a module.

Thanks!

ezyang commented:
What version of PyTorch are you running?

Excerpts from Javier Zazo's message of 2019-07-05 02:41:34 -0700:
> jzazo commented on this pull request.
> 
> >  
>      @weak_script_method
>      def forward(self, input):
> +        if self.padding_mode == 'circular':
> +            expanded_padding = ((self.padding[0] + 1) // 2, self.padding[0] // 2)
> +            return F.conv1d(F.pad(input, expanded_padding, mode='circular'),
> +                            self.weight, self.bias, self.stride,
> +                            _single(0), self.dilation, self.groups)
>          return F.conv1d(input, self.weight, self.bias, self.stride,
> 
> In the documentation it appears padding_mode is supported, but when I try to use it tells me it is an unexpected argument. I will do the manual padding, as in the module, but it would be convenient if the operation is supported.
> In my case, my filter is a tensor, not a parameter, and I would prefer not to use a module.
> 
> Thanks!
>

jzazo commented:
I am running 1.1.0. Thanks!

thomasahle commented:
If I try to run
`F.pad(torch.arange(4), (0,1), mode='circular')`
or
`F.pad(torch.arange(9).reshape(3,3), (0,1,1,0), mode='circular')`
I get the error
`NotImplementedError: Only 2D, 3D, 4D, 5D padding with non-constant padding are supported for now`.
(Torch v 1.11)
Am I doing this wrong?
